### ISSUE TITLE
Opt in for more TS/JS-friendly enum representation.

### ISSUE THREAD

**Issue by back2dos** [HaxeFoundation/haxe#8591]
It would be nice if there were a way to make the compiler emit code that puts the constructor name into the object (under e.g. a `$type` property), because that way we could use enums in APIs we expose to other frontend teams (which use JS/TS). Maybe via a compiler flag, that would also control the property name.

**Comment by nadako** [comment]
Somewhat related to #7165 (still has to rename _hx_index to `$index` etc). We actually have a define for using name instead of index, but IIRC that's mostly a debugging feature.

**Comment by fullofcaffeine** [comment]
Yes, please. I just stumbled upon this limitation when using `hxtsdgen` to directly expose a `coconut.data` model to TS (it failed when trying to generate `tink.core.Noise`). See: https://github.com/nadako/hxtsdgen/issues/28.

**Comment by benmerckx** [comment]
How about a reference to the constructor/self? Would allow js side to do:
````js
switch (value.$type) {
  case Option.Some:
    console.log(value.v)
    break
  case Option.None:
    console.log('none')
    break
}
````
Edit: just realized this wouldn't fly in Typescript because it needs a constant type to switch over

**Comment by nadako** [comment]
> Edit: just realized this wouldn't fly in Typescript because it needs a constant type to switch over

This [seems to work](https://www.typescriptlang.org/play?#code/DYUwLgBAJAXAhgOwJ4G4CwAoTBjA9ggZ0gAs4APEAfQBMDKB5ABzAEt8IBeCAb0wEgAcvhAwIACiiceUMEkYiIAN1wtqEAAwBfADTQAdDLkgpUXVACU2-gGVcAWwUSpAMwCuCbK3xjF8ZOZ4IACdwVyCEaVl5UVMlUUVNCB1oc0xNTExQSDIpUgoaOiYvBD1bBzEARgBWc3QMAgB3FjBsYnEyAyiQAN4MPmw4AmM8qloGZjYSoQQRfn78AlxQPWBcAHMxACJpkE3auYAjELgAazr+weHyUcKJ-FL7Wb75wiWQFfWtst3dDsV955HECnOrpDBAA).

```ts
let $:any;

const haxe_ds_Option = {
	None: ($ = {$type: void 0}, $.$type = $, $),
	Some: ($ = function(v:any) { return {$type: $, v: v} }, $)
}

let x = haxe_ds_Option.Some(15);
switch (x.$type) {
	case haxe_ds_Option.None:
		console.log("None");
		break;
	case haxe_ds_Option.Some:
		console.log("Some", x.v);
		break;
}
```

It doesn't provide any type-safety on TS side though, and I'm not sure it's even possible to have a proper TS "extern" for that without using literal integers/strings.

**Comment by benmerckx** [comment]
Exactly, I meant Typescript won't be able to distinguish types this way - a constant (or literal) integer or string type would be needed in addition to typing it as a union. Although you can trick it somewhat:

```typescript
let $:any;

type haxe_ds_Option<T> =
	| {$type: 0}
	| {$type: 1, v: T}

const haxe_ds_Option: {
	None: 0 & haxe_ds_Option<any>,
	Some: 1 & (<T>(v: T) => haxe_ds_Option<T>)
} = {
	None: ($ = {$type: void 0}, $.$type = $, $),
	Some: ($ = function(v:any) { return {$type: $, v: v} }, $)
}

let x = haxe_ds_Option.Some(15);
switch (x.$type) {
	case haxe_ds_Option.None:
		console.log("None");
		break;
	case haxe_ds_Option.Some:
		console.log("Some", x.v);
		break;
}
```

[Playground Link](https://www.typescriptlang.org/play?ssl=23&ssc=2&pln=1&pc=1#code/DYUwLgBAJAXAhgOwJ4G4CwAoTYkAcQQAWcAHiAPoAmAzuQPK5gCWA9ggDwAqAfBALyYAkAB8IAbyg58MCAAYAvkNESpIGQEYANBABuMzoqwYAxm2qRiZKrQbM2MsUIBybNXIgAyIqQo16jVg5EJG5NIQBlFgBbN3VPCAAKLm4EvQhOAEp+XktfGwC2ZIzMeX5xZ1cZBKgylTw3HRYmSjl5bSgAOkl6sqh2jLCMQUiYqpq+CAAzAFcEYzsEVPhkLLEIACdwafWEcW7paG00nVK26GKMQ0xQSBIy3Ot-BY6RkAT1AFYM9AxqAHcmGBjIREiQuqpVkJjHBqAQHn5bIEOi4EGohIJTAhqCxQB1gCwAOYJABEKJAxO+6IARps4ABrH4YmFwnyPRFsF7RNFDDFmHEgPGEkmvYnaME6Sk8mkgek-QxAA)

I guess it's something to solve for external d.ts generators.
